### DISCUSSION TITLE
User Mode Linux (Termux-on-Termux)

### DISCUSSION THREAD

**Discussion by lars18th** [termux/termux-packages#8555] [Ideas]
Hi,

I want to share with you a new proposal. The background of this is based on the large work done to execute Docker inside Android/Termux. However, from my point of view this effort requires to continue until we can run Docker on NON-ROOT devices.

Let me to explain how to achieve this:

- Using the User Model Linux (aka UML) we can execute a Linux kernel as a process. See the recent documentation on the Linux kernel about this: https://www.kernel.org/doc/html/latest/virt/uml/user_mode_linux_howto_v2.html#introduction (and note the recent changes in the kernel tree regarding the UML virtualization). Even you think this could be slow, in fact, it only has three drawbacks: performance is limited to only one processor, all syscalls are slow (then process and threads generate overhead), and I/O could be slow (if the networking can't be handled in the host kernel). Despite these problems the performance is nearly similar to hardware accelerated virtualization. Then why not use it?

The first goal to achieve this (the idea of running a Termux environment inside a UML kernel) could be done on three steps:

1. Create a new "termux-root-pakage" called "uml-linux" that generates a UML kernel that can run inside Termux in rooted devices.
2. Based on the previous package a second one will be "termux-on-uml". This second packages uses the "uml-linux" package to setup a new Termux environment completly agnostic of the host Android device. The objective of this package is provide a clean Termux environment with the mininum to run it. And perhaps some glue to connect some host API calls to the guest environment.
3. Finally, the most complex task will be the creation of the termux package "uml-linux-nonroot". The challenge with this package is to provide all the required functionalities using only the user restricted Android Linux kernel and the Android API. For example, for networking a new BESS module requires to be developed perhaps using the Android API for VPNs.

You can imagine that this will open a large interesting oportunities. For example (but not limited to):
- Running Docker in any non-rooted Android device.
- Install a Termux environment using virtual disk (images) stored in the SD card wit or without root privileges (for rooted devices this will interesting too for easy and compact installation).

You can read some interesting efforts about this technology on these links:
- Example of UML in recent Linux x86 (only to understand how it works, as recent kernels have superseded the slipr part): https://christine.website/blog/howto-usermode-linux-2019-07-07
- Docker in Android literature (about how to execute containers in Android devices with kernel support, that will could be overcomed when executed inside the UML Linux): https://github.com/CGCL-codes/Android-Container
- Docker on Termux discussion (to understand the efforts done): https://gist.github.com/arno01/ebf570af208e28c1a0cf78da4f63bc9c

I hope someone feels that this idea is feasible and interesting.
Please, comment your point of view.
Regards.

**Comment by ghost** [comment]
Does UML have proper support for arm(64) somehow? checked the kernel sources and it seems it only supports x86(64) architecture

**Reply by ghost** [reply]
Yes, it is x86(64) only and no one of Termux devs can/will port it.

**Reply by ghost** [reply]
There are few more issues:
* UML is available only for x86 & x86_64 and can't be run under QEMU user either.
* Running UML requires certain things that normally available under `proot`, e.g. SysV IPC.

I've already tried it. Idea interesting, but unlikely can be accomplished. As said previously: porting UML is wontfix. We don't do any kernel-side development.

**Reply by lars18th** [reply]
Hi @xeffyr ,

I'm not aware of the x86 (32/64 bit) support _only_ of the UML kernel. Sorry for not discover this before. Anyway, please note that _using_ UML is not a _kernel_side_ development. UML is a technology to run a Linux kernel as a process in the userland. Then you don't need to touch in any way the host device kernel.

Futhermore, I feel that at some point in the future the UML support of the Linux kernel will be updated to support the ARM platform. You only need to see that the development of this module was reactivated recently. And for sure some ARM kernel mantainers will want to have this technology available. Another question is when it will be included. And for sure if this UML kernel could be then running inside a regular Android device. But if in the future this will be supported... then you won't like to use it to ehance the Termux support?

**Reply by lars18th** [reply]
Hi,

Here are the instructions to port the UML kernel to other platforms: http://user-mode-linux.sourceforge.net/old/arch-port.html

I know that this is not a job for the Termux community. However, I post here if someone wants to start the work for running it into a Raspbian machine.
